### PR TITLE
Customized error raised in Python model needs be passed back to client

### PR DESCRIPTION
**Pull request by lennon310**
(1) Change `IOException` to explicit `InvalidProtocolBufferException`.

(2) Avoid re-throwing the `APIException` thrown when `httpResponse.getStatusCode()` is non-2xx. However, any other exception is still caught and thrown as `APIExeption`.

(3) Directly throw the exception with the payload as its message. The payload will be the `dict` defined in user defined exception in Python model.

(4) The exception wrapper (`ExceptionControllerAdvice`class) should create `SeldonMessage` instead of `Status` (bug fix).

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **cliveseldon**

If they are not already assigned, you can assign the PR to them by writing `/assign @cliveseldon` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Hi @lennon310. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) or [seldonio](https://github.com/orgs/seldonio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/seldonio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java [hunk 1]
```diff
@@ -491,19 +490,26 @@ private SeldonMessage queryREST(
         ResponseEntity<String> httpResponse =
             restTemplate.postForEntity(uri, request, String.class);
         try {
+          SeldonMessage.Builder builder = SeldonMessage.newBuilder();
+          String response = httpResponse.getBody();
+          logger.debug(response);
           if (httpResponse.getStatusCode().is2xxSuccessful()) {
-            SeldonMessage.Builder builder = SeldonMessage.newBuilder();
```
Comment (ahutterTA) [inline engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: revert the change to move builder outside of this block?

File: engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java [hunk 2]
```diff
@@ -513,14 +519,10 @@ private SeldonMessage queryREST(
         }
       } catch (ResourceAccessException e) {
         logger.warn("Caught resource access exception ", e);
-      } catch (IOException e) {
-        logger.error("Couldn't retrieve prediction from external prediction server - ", e);
-        throw new APIException(
-            APIException.ApiExceptionType.ENGINE_MICROSERVICE_ERROR, e.toString());
-      } catch (Exception e) {
```
Comment (ahutterTA) [inline engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: Do you still need this catch all case?

**Comment by lennon310** [comment]
I will add a test on this API

File: engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java [hunk 3]
```diff
@@ -491,19 +491,30 @@ private SeldonMessage queryREST(
...
+            if (seldonMessageStatus == null) {
+                throw new APIException(
+                        APIException.ApiExceptionType.ENGINE_MICROSERVICE_ERROR,
+                        String.format("Bad return code %d", httpResponse.getStatusCodeValue()));
+            }
+            else
+            {
+                APIException.ApiExceptionType exceptionType =
+                        APIException.ApiExceptionType.CUSTOMIZED_EXCEPTION;
+                exceptionType.setMessage(seldonMessageStatus.getReason());
```
Comment (adriangonz) [inline engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: Since it's changing the message of the global `APIException.ApiExceptionType.CUSTOMIZED_EXCEPTION`, is this thread-safe? 

What would happen with a different thread using `APIException.ApiExceptionType.CUSTOMIZED_EXCEPTION`? Which message would it get?

Comment (lennon310) [reply engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: Good point, I got rid of the customized enum, and use a constructor instead to avoid such a concern.

Comment (adriangonz) [reply engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: I'm not too familiar with Java's `enum`, but is it possible to create your own at runtime? For example, keeping a similar structure as how it was before the last commit, but doing instead:

```java
APIException.ApiExceptionType exceptionType =
    APIException.ApiExceptionType(CUSTOM_EXCEPTION, seldonMessageStatus.getReason(), 400);
throw new APIException(exceptionType, seldonMessageStatus.getInfo());
```

That way we would reduce the changes to the `APIException` class. I'm a bit wary that adding the new fields which were previously part of `APIException.ApiExceptionType` may be duplicating that functionality.

Comment (lennon310) [reply engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: I doubt we can,  since enum is static, and its constructor has to be private.

Comment (adriangonz) [reply engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java]: I agree. 

Adding an inherited `CustomException` from `APIException` could be a third option, but it doesn't add much value. It's something that can be done in a future if needed.

BTW I really like the new simplified interface of `APIException` (without the `getExceptionType()` calls)! It also makes it easier to extend it in a future.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/1010#issuecomment-547851661" title="Approved">adriangonz</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [adriangonz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
